### ISSUE TITLE
light.setColor callback very slow

### ISSUE THREAD

**Issue by RalphHogenbirk** [jangxx/node-magichome#11]
For controlling my LED strips at home, I'm calling turnOff, turnOn, setColor and setColorWithBrightness functions. turnOff and turnOn finish quite fast (~3 sec), but setColor hangs for  about 15 seconds before executing the callback, even though the light successfully changed color within the first 2-3 seconds.

Any way to improve this behaviour? The domoticz setup I'm using to execute the commands is not happy with scripts hanging for more than 10 seconds.

**Comment by jangxx** [comment]
I'm trying to think of a way this (or any) library could influence the controllers in a way that would solve your problem, but I have a hard time coming up with one.

Frist, let me get this straight; You have a controller, which confirms every request (so `characteristics.wait_for_reply` is `true`), and you're having a problem where the controller sends it's reply back with a huge delay.

The naive solution would probably be to just set `wait_for_reply` to `false`, which calls the callback immediately after the command is sent out. I'm pretty sure that this would introduce more problems than it would solve however. First, the controller now runs into problems trying to send it's confirmation through a closed socket. Secondly, even if the controller manages to send it's reply, the library is now in a position where it receives unexpected data (the reply we did not handle previously) which will probably mess some things up since this case is currently not handled explicitly.

Have you checked your network conditions? You said that `turnOn` and `turnOff` finish "quite fast" in about 3 seconds, but this seems very long to me. The two controllers I have (one which confirms commands and one which doesn't) usually react within milliseconds. Occasionally, the non-confirming controller takes several seconds to react to an issued command, but since the commands are not confirmed anyway, this leads to no problems.

Since this library can't influence the speed at which a controller replies, you could wrap the function call with some sort of timeout yourself and then just hope that the controller does not return an error-indicating reply (which it usually doesn't do). The code might look something like this:

```javascript
let control = new Control(...);

function setColorWithTimeout(red, green, blue, timeout_millis) {
	return new Promise((resolve, reject) => {
		let resolved = false;

		let timeout = setTimeout(() => {
			resolved = true;
			resolve();
		}, timeout_millis);

		control.setColor(red, green, blue, (err, success) => {
			if (!resolved) {
				clearTimeout(timeout);
				if (err || !success) return reject(err);
				return resolve(err);
			} // otherwise just discard the late reply quietly
		});
	});
}
```

I'm pretty sure there are libraries for this as well, which you could use to have custom timeouts for all of the control methods without much work.

If you have any other ideas for the root of your long delays please let me know, but I'm pretty sure that these stem from either your network or the controller itself, both of which this library can't do anything about.

**Comment by RalphHogenbirk** [comment]
Thanks for the quick response and all the info @jangxx! I'll try to find some time one of these evenings to dive into this. Actually some of the issue might be in the old raspi I'm running this on. But, my domoticz setup on there works fine and is quite responsive, so you would say the raspi hardware should be good. The magic-home app on my android phone has virtually no delay with the same led strip, which excludes the controller and wifi connection. 

In any case, there is a lot of delay difference between the on/off commands and the setColor command, which is curious. Any pointers on major differences between the two commands? Maybe if I find the source for this difference, I can track it down to the root cause of all my delay issues.

**Comment by RalphHogenbirk** [comment]
Thanks! I was already testing the scripts without domoticz, only wrote a simple wrapper javascript file that could be called from domoticz later on, but I was only using it from the commandline so far. I just mentioned the domoticz setup to indicate that my raspi isn't super slow or without memory or whatever :)

I'll test with type2 as you mentioned, try running it from my laptop and will test the python tool as a last option. Thanks for helping out so far!

**Comment by jangxx** [comment]
Hey, I had some time this evening and finally got around to rewriting this library with native Promise support and also the missing CustomMode stuff. I don't really expect that your problem is fixed now, but maybe some problem I overlooked before is now simply lost to the rewrite (or I introduced more bugs, which is more likely I guess). Maybe have another go with version 2.0.0 and tell me if you problem persists.
